Ext4Pkg/Ext4Dxe/File.c

```diff
@@ -134,13 +134,222 @@ Ext4DirCanLookup (
   return (File->Inode->i_mode & EXT4_INO_PERM_EXEC_OWNER) == EXT4_INO_PERM_EXEC_OWNER;
 }
 
+/**
+  Reads a fast symlink file.
+
+  @param[in]      Partition   Pointer to the ext4 partition.
+  @param[in]      File        Pointer to the open symlink file.
+  @param[out]     AsciiSymlink     Pointer to the output ascii symlink string.
+  @param[out]     AsciiSymlinkSize Pointer to the output ascii symlink string length.
+
+  @retval EFI_SUCCESS           Fast symlink was read.
+  @retval EFI_OUT_OF_RESOURCES  Memory allocation error.
+**/
+STATIC
+EFI_STATUS
+Ext4ReadFastSymlink (
+  IN     EXT4_PARTITION  *Partition,
+  IN     EXT4_FILE       *File,
+  OUT    CHAR8           **AsciiSymlink,
+  OUT    UINTN           *AsciiSymlinkSize
+  )
+{
+  EFI_STATUS  Status;
+  CHAR8       *AsciiSymlinkTmp;
+
+  AsciiSymlinkTmp = AllocatePool (EXT4_FAST_SYMLINK_SIZE + 1);
+  if (AsciiSymlinkTmp == NULL) {
+    Status = EFI_OUT_OF_RESOURCES;
+    DEBUG ((DEBUG_FS, "[ext4] Failed to allocate symlink ascii string buffer\n"));
+    return Status;
+  }
+
+  CopyMem (AsciiSymlinkTmp, File->Inode->i_data, EXT4_FAST_SYMLINK_SIZE);
+
+  //
+  // Add null-terminator
+  //
+  AsciiSymlinkTmp[EXT4_FAST_SYMLINK_SIZE] = '\0';
+
+  *AsciiSymlink = AsciiSymlinkTmp;
+  *AsciiSymlinkSize = EXT4_FAST_SYMLINK_SIZE + 1;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Reads a slow symlink file.
+
+  @param[in]      Partition        Pointer to the ext4 partition.
+  @param[in]      File             Pointer to the open symlink file.
+  @param[out]     AsciiSymlink     Pointer to the output ascii symlink string.
+  @param[out]     AsciiSymlinkSize Pointer to the output ascii symlink string length.
+
+  @retval EFI_SUCCESS           Slow symlink was read.
+  @retval EFI_OUT_OF_RESOURCES  Memory allocation error.
+  @retval EFI_INVALID_PARAMETER Slow symlink path has incorrect length
+  @retval EFI_VOLUME_CORRUPTED  Symlink read block size differ from inode value
+**/
+STATIC
+EFI_STATUS
+Ext4ReadSlowSymlink (
+  IN     EXT4_PARTITION  *Partition,
+  IN     EXT4_FILE       *File,
+  OUT    CHAR8           **AsciiSymlink,
+  OUT    UINTN           *AsciiSymlinkSize

```

Comment:
Tbh, this could be `UINT32` (same for the Slow function). `UINTN` is nasty. :)

Reply:
Maybe UINT64? Then we close question with UINTN cast below

Reply:
Why UINT64? This is bounded. A symlink shouldn't exceed 4 GB. :D

Reply:
Thats why comparison with EFI_PATH_MAX below. Ok, I got your point, we can make SymlinkSize UINT32, but we will have another cast to UINT32 in its caller, so we need to make check in Ext4ReadSymlink, but we have Ext4ReadSlowSymlink, in which  we take SymlinkSize dynamically walking by blocks and it can return > 32bit SymlinkSize in bad situation. Your ideas how do it right?

Reply:
> but we have Ext4ReadSlowSymlink, in which we take SymlinkSize dynamically walking by blocks and it can return > 32bit SymlinkSize in bad situation

Why? as you said yourself, it is bounded by `EFI_PATH_MAX <= MAX_UINT32`. The caller shouldn't need a cast, it just uses `UINT32` throughout.

Reply:
Also in such case we will put UINT32 into UINTN in Ext4Read.
`Status = Ext4Read (Partition, File, SymlinkTmp, File->Position, &SymlinkSizeTmp);`

Reply:
I think solved in https://github.com/acidanthera/audk/commit/2e312048ee078d67484bb73f12bf0466dcf615c6